contracts/System.sol

```diff
@@ -80,4 +122,171 @@ contract System {
+  }
+
+  function _ext() private pure returns (ExtStorage storage _extStorage) {
+    // @dev create 'floating' storage unit that will be positioned at _offset rather sequentially at the end of the contract
+    // and thus will not collide with storage of derived contracts
+    // see https://solidity.readthedocs.io/en/v0.8.4/internals/layout_in_storage.html#layout-of-state-variables-in-storage
+    bytes32 _offset = keccak256("core.system.extStorage");
+    assembly { _extStorage.slot := _offset }
+  }
+
+  function _validatorSet() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.validatorSet);
+    } else {
+      return _VALIDATOR_CONTRACT_ADDR;   
+    }
+  }
+
+  function _slash() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.slash);
+    } else {
+      return _SLASH_CONTRACT_ADDR;   
+    }
+  }
+
+  function _systemReward() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.systemReward);
+    } else {
+      return _SYSTEM_REWARD_ADDR;   
+    }
+  }
+
+  function _lightClient() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.lightClient);
+    } else {
+      return _LIGHT_CLIENT_ADDR;   
+    }
+  }
+
+  function _relayerHub() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.relayerHub);
+    } else {
+      return _RELAYER_HUB_ADDR;   
+    }
+  }
+
+  function _candidateHub() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.candidateHub);
+    } else {
+      return _CANDIDATE_HUB_ADDR;   
+    }
+  }
+
+  function _govHub() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.govHub);
+    } else {
+      return _GOV_HUB_ADDR;   
+    }
+  }
+
+  function _pledgeAgent() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.pledgeAgent);
+    } else {
+      return _PLEDGE_AGENT_ADDR;   
+    }
+  }
+
+  function _burn() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.burn);
+    } else {
+      return _BURN_ADDR;   
+    }
+  }
+
+  function _foundation() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.foundation);
+    } else {
+      return _FOUNDATION_ADDR;   
+    }
+  }
+
+  function updateContractAddr(address validatorSet_, address slash_, address systemReward_, 
+                              address lightClient_, address relayerHub_, address candidateHub_, 
+                              address govHub_, address pledgeAgent_, address burn_, address foundation_) 
+           external onlyLocalTestMode onlyTestModeDeployer updateContractAddrCalledOnce {
```

Comment:
placing updateContractAddr() in the production contract is highly risky. To fire-proof it I use 3 layers of 
protection;
1. constructor must be invoked else onlyTestModeDeployer will revert  [will never happen in production]
2. verify that chainid is not mainnet else onlyLocalTestMode will fail  [will never happen in production]
3. verifies that the function can never be called more than once, which is probably redundant but that's ok

Reply:
If someone clones this codebase for use as the genesis contracts of a new blockchain, they would be exposed to the risk, correct?

Just flagging it as something we (and external folks who aren't as familiar with the codebase) want to be mindful of

Reply:
correct! this issue must be documented and made clear to any future clone. 
also note that the additional protection I have added - especially that of onlyDeployer protects us from hacking even in such a case

Reply:
yep that's the reason I implemented _isLocalTestNode() with NEQ logic